### PR TITLE
#9 Task Dzmitry

### PR DESCRIPTION
**Pull request by marioti95**
https://travis-ci.com/github/marioti95/js-assignments/builds/174502100

### CODE REVIEW

File: task/09-functions-n-closures-tasks.js [hunk 1]
```diff
@@ -25,11 +24,13 @@
...
  *
  */
-function getComposition(f,g) {
-    throw new Error('Not implemented');
+function getComposition(f, g) {
+  function inner(x) {
+    return f(g(x));
+  }
+  return inner;
 }
```
Comment (FedorovD) [inline task/09-functions-n-closures-tasks.js:29]: why not do just `return x => f(g(x))`  ?

File: task/09-functions-n-closures-tasks.js [hunk 2]
```diff
@@ -47,10 +48,12 @@ function getComposition(f,g) {
  *
  */
 function getPowerFunction(exponent) {
-    throw new Error('Not implemented');
+  function exp(x) {
+    return Math.pow(x, exponent);
+  }
+  return exp;
```
Comment (FedorovD) [inline task/09-functions-n-closures-tasks.js]: same

File: task/09-functions-n-closures-tasks.js [hunk 3]
```diff
@@ -104,10 +127,20 @@ function memoize(func) {
...
+      try {
+        return func();
+      } catch (err) {
+        if (i === attempts) {
+          throw err;
+        }
+      }
+    }
+  }
+  return try_func;
```
Comment (FedorovD) [inline task/09-functions-n-closures-tasks.js]: unnecessary extra code, 
use just `return () => {...}`

File: task/09-functions-n-closures-tasks.js [hunk 4]
```diff
@@ -132,10 +165,16 @@ function retry(func, attempts) {
...
 function logger(func, logFunc) {
-    throw new Error('Not implemented');
+  function log() {
+    let str = `${func.name}(${JSON.stringify([...arguments]).slice(1, -1)})`;
+    logFunc(str + ' starts');
+    let value = func.apply(null, arguments);
+    logFunc(str + ' ends');
+    return value;
+  }
+  return log;
```
Comment (FedorovD) [inline task/09-functions-n-closures-tasks.js:166]: same

File: task/09-functions-n-closures-tasks.js [hunk 5]
```diff
@@ -150,10 +189,15 @@ function logger(func, logFunc) {
...
  */
 function partialUsingArguments(fn) {
-    throw new Error('Not implemented');
+  let outer = Object.values(arguments);
+  outer.shift();
+  function partArgs() {
+    let inner = Object.values(arguments);
+    return fn(...outer, ...inner);
+  }
+  return partArgs;
```
Comment (FedorovD) [inline task/09-functions-n-closures-tasks.js:189]: the same

**Comment by FedorovD** [comment]
Also fix code formatting, return to original one.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/09-functions-n-closures-tasks.js (hunk 5 above)
Comment (marioti95) [reply task/09-functions-n-closures-tasks.js:189]: I am using inner function clouser hear

Comment (marioti95) [reply task/09-functions-n-closures-tasks.js:189]: I am using inner function clouser here

**Comment by FedorovD** [comment]
Accepted.

**Comment by FedorovD** [comment]
@marioti95 whats your slack name?

**Comment by marioti95** [comment]
> @marioti95 whats your slack name?

Dzmitry

**Comment by marioti95** [comment]
Dzmitry

ср, 29 июл. 2020 г. в 12:06, Danil <notifications@github.com>:

**Pull request closed without merging**